### PR TITLE
DrivAerML: Transfer AirfRANS grad-clip=1.5 breakthrough to 4L/320d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS just achieved a 26.4% improvement (0.01271 → 0.00935) by increasing grad-clip from 1.0 to 1.5 (PR #2737). The hypothesis is that gradient instability is similarly limiting DrivAerML, and that the same moderate clipping threshold will unlock a significant improvement on the 4L/320d winning architecture.

Key reasoning:
- On AirfRANS, clip=1.5 outperformed both clip=0.5 (too tight) and clip=1.0 (still too tight) — moderate clipping allows enough gradient magnitude for effective learning while suppressing destructive spikes at cosine LR peaks
- DrivAerML also uses cosine annealing with T_max=30, creating the same periodic spike pattern at LR peaks
- The 4L/320d architecture already has the throughput advantage (257 epochs vs 151 for 4L/384d in 180 min), giving more opportunities to benefit from stable gradient flow
- If the improvement transfers even partially, this could push DrivAerML well below 5.027%

## Instructions

No code changes needed. Run the following command exactly — it applies grad-clip=1.5 to the current winning 4L/320d configuration:

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.5 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 320 \
  --model-heads 5 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-4L320d-gc15
```

**CRITICAL:** The `SENPAI_TIMEOUT_MINUTES=180` and `SENPAI_MAX_EPOCHS=9999` env vars must be included. Do not modify any other hyperparameters — this is a single-variable test of grad-clip=1.5.

## Baseline

- **Primary metric:** `val_primary/surface_rel_l2_pct` (lower is better)
- **Current best:** 5.027% (val) / 6.244% (test) at epoch 257
- **Best PR:** #2648 (zoro — 4L/320d/5H, T_max=30, lr=5e-4, no-EMA, Fourier, 257 epochs in 180-min)
- **External target:** <3.71% (AB-UPT, ~500 epochs) — 1.35x gap remaining
- **AirfRANS analog:** grad-clip=1.5 gave 26.4% improvement on AirfRANS (PR #2737)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 320 --model-heads 5 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

## Expected outcome

If gradient instability is a shared bottleneck, expect 10-26% improvement in `val_primary/surface_rel_l2_pct`. Beat target: < 5.027%. Even a 5% relative improvement (< 4.78%) would compound meaningfully with future changes.

Report the W&B run ID, best val metric, epoch of best, and number of total epochs completed in 180 minutes.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT taki:

## Results

### Key Metrics

| Metric | Baseline | This Run | Change |
|--------|----------|----------|--------|
| `val_primary/surface_rel_l2_pct` | 5.027% | **5.207%** (epoch 245) | +3.6% (worse) |
| `test_primary/surface_rel_l2_pct` | 6.244% | **6.800%** | +8.9% (worse) |
| Epochs completed | 257 | 267 | +10 |

### Experiment Details

- **W&B run ID:** `pusx1d2d` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/pusx1d2d))
- **Peak memory:** Not captured in logs (GPU released post-training)
- **Command used:**
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.5 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group drivaerml-4L320d-gc15 --agent taki --wandb-name "taki/drivaerml-4L320d-gc15"
```

### What happened

**The grad-clip=1.5 hypothesis did not beat the baseline on DrivAerML.** Best val was 5.207% vs the 5.027% target, a 3.6% relative deficit.

The training showed excellent convergence behavior:
- Grad clip events dropped from 330/epoch (epoch 1) to ~0-2/epoch (late training), indicating gradients naturally stabilized
- Consistent improvement at cosine cycle troughs (every 15 epochs), with val dropping from 57.55% → 5.21% over 267 epochs
- The model plateaued in the 5.2-6.0% range from epoch ~215 onwards, oscillating with cosine LR cycles but not breaking below 5.21%

**Why it didn't transfer like AirfRANS:** The AirfRANS improvement (26.4%) came from a regime where gradient instability was a binding constraint. On DrivAerML, the grad norms were already well-behaved — by epoch ~50, nearly zero clips per epoch, meaning the clip threshold of 1.5 was rarely active. The baseline (presumably clip=1.0 or no clip) likely wasn't suffering from destructive gradient spikes on DrivAerML, so loosening the clip provided no benefit and may have allowed slightly more noise.

**Key observation:** The 267 epochs completed (vs 257 baseline) with similar throughput, suggesting grad-clip=1.5 doesn't add meaningful compute overhead. The test metric (6.80%) was notably worse than val (5.21%), indicating some overfitting in the later epochs.

### Suggested follow-ups

1. **Try grad-clip=1.0 explicitly** if the baseline doesn't use clipping — moderate clipping might help more than loosening it
2. **Weight decay tuning** — the val/test gap suggests regularization could help
3. **Longer T_max** (e.g., 60) — slower cosine cycles might help the model settle better at this maturity
4. **EMA** — the no-EMA setting might be leaving test performance on the table given the val/test gap

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2749 because ADVISOR: Closing — gc=1.5 at 4L/320d reached 5.207% vs baseline 4.619% at 4L/512d. Two issues: (1) gc=1.5 provided no benefit — DrivAerML gradients are naturally well-behaved by epoch 50, so loosening the clip just adds noise. (2) Architecture superseded by 4L/512d baseline. Good insight that gc is not the bottleneck for DrivAerML..
